### PR TITLE
Support casts in constants

### PR DESCRIPTION
**Pull request by neurosie**
Addresses #521.

### CODE REVIEW

File: src/bindgen/ir/constant.rs [hunk 1]
```diff
@@ -45,7 +49,8 @@ impl Literal {
             Literal::PostfixUnaryOp { .. }
             | Literal::BinOp { .. }
             | Literal::Expr(..)
-            | Literal::Path(..) => {}
+            | Literal::Path(..)
+            | Literal::Cast { .. } => {}
```
Comment (emilio) [inline src/bindgen/ir/constant.rs:53]: This needs to call `replace_self_with` in the subexpression, but this is also an issue with other literals.

File: src/bindgen/ir/constant.rs [hunk 2]
```diff
@@ -104,6 +110,9 @@ impl Literal {
                 right.rename_for_config(config);
             }
             Literal::Expr(_) => {}
+            Literal::Cast { ref mut value, .. } => {
```
Comment (emilio) [inline src/bindgen/ir/constant.rs:113]: The type should also be renamed to account for the configuration. This should be trivial to test.

File: src/bindgen/ir/constant.rs [hunk 3]
```diff
@@ -252,6 +274,12 @@ impl Literal {
                 right.write(config, out);
                 write!(out, ")");
             }
+            Literal::Cast { ref ty, ref value } => {
```
Comment (emilio) [inline src/bindgen/ir/constant.rs:277]: We should probably wrap the whole expression in parens.

**Comment by neurosie** [comment]
Something like `const FOO: *mut Struct = 0 as usize as *mut Struct;` is not currently handled, because it's not a primitive or pointer to one (decided in code [here](https://github.com/eqrion/cbindgen/blob/5c7a7fca4233c7cb73038c58599cb1938c8219ad/src/bindgen/ir/constant.rs#L303)).

However, I did find a problem with something like this:
```
struct A;
pub const B: i32 = 0 as *const A as i32;
```

This gets reordered to
```
static const int32_t B = (int32_t)(const A*)0;
struct A;
```
which fails to compile. It gets reordered because [primitive or pointer to primitive consts are always printed first](https://github.com/eqrion/cbindgen/blob/5c7a7fca4233c7cb73038c58599cb1938c8219ad/src/bindgen/bindings.rs#L218-L224). This is kind of a weird edge case, do you think it's worth handling?

**Comment by emilio** [comment]
Yes, we should handle it somehow, either bailing out or handling it correctly. Seems like "is a primitive or pointer to it" is not enough for these. Probably we really want something like "uses non-primitive types" (where pointer is considered a "primitive" type), and check that.

File: src/bindgen/ir/constant.rs (hunk 3 above)
Comment (neurosie) [reply src/bindgen/ir/constant.rs:312]: Casts have pretty high precedence (https://en.cppreference.com/w/c/language/operator_precedence), so I don't think that's necessary. I can add it if you really want though.

File: src/bindgen/ir/constant.rs [hunk 4]
```diff
@@ -72,6 +91,24 @@ impl Literal {
...
+    pub fn uses_only_primitive_types(&self) -> bool {
+        match self {
+            Literal::Expr(..) => true,
+            Literal::Path(..) => true,
+            Literal::PostfixUnaryOp { ref value, .. } => value.uses_only_primitive_types(),
+            Literal::BinOp {
+                ref left,
+                ref right,
+                ..
+            } => left.uses_only_primitive_types() & right.uses_only_primitive_types(),
```
Comment (emilio) [inline src/bindgen/ir/constant.rs:107]: Any reason this is using `&` instead of `&&`?

File: src/bindgen/ir/constant.rs (hunk 3 above)
Comment (emilio) [reply src/bindgen/ir/constant.rs:312]: Macros that expand to complex stuff without parens always make me a bit nervous, but you're probably right, we can always change this if needed.

File: src/bindgen/ir/constant.rs (hunk 4 above)
Comment (emilio) [reply src/bindgen/ir/constant.rs:107]: I assume not, will fix this as a follow-up.

**Pull request merged**